Data-Structures/Linked-List/CycleDetectionII.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * A LinkedList based solution for Detecting a start node of the Cycle in a list.
+ * https://en.wikipedia.org/wiki/Cycle_detection
+ */
+
+function detectCycleNode(head) {
+  /*
+  Problem Statement:
+  Given the head of a linked list, return the node where the cycle begins. If there is no cycle, return null.
+  Link for the Problem: https://leetcode.com/problems/linked-list-cycle-ii/
+  */
+
+  let length = 0
+  let fast = head
+  let slow = head
+
+  while (fast !== null && fast.next !== null) {
+    fast = fast.next.next
+    slow = slow.next
+    if (fast === slow) {
+      // Calculate length of the cycle
+      length = lengthCycle(slow)
+      break
+    }
+  }
+
+  if (length === 0) {
+    // If length = 0, return null
+    return null
+  }
+
+  //Point the nodes at head
+  fast = head
+  slow = head
+  // Move slow pointer ahead by 'length' of cycle times
```

Comment:
The purpose of this is unclear to me. If `length` is in nodes, then `fast = slow = head` after this and this is effectively dead code. If `length` is in "edges" / "links" (one fewer than nodes), then you will get the predecessor of `head` in the cycle, but what for?

Reply:
Here, the 'length' refers to the number of nodes to move forward the 'slow' pointer in the linked list. (while keeping the 'fast' pointer at head only ). Then we'll move both pointers one by one to get the starting node of cycle. I have attached an image to help illustrate this point and my thought process.

Reply:
Thought Process: 
![CycleNode1](https://github.com/TheAlgorithms/JavaScript/assets/118964099/78544105-84ac-43c0-84a2-c45bd195a664)

Reply:
![CycleNode2](https://github.com/TheAlgorithms/JavaScript/assets/118964099/a04ec2ea-b7d2-4ca9-8a66-6daabcb14b2a)